### DISCUSSION TITLE
Is it possible to handle headers with `__float128`?

### DISCUSSION THREAD

**Discussion by amontoison** [JuliaInterop/Clang.jl#508] [General]
Some Fortran libraries, like `CUTEst`, can be easily compiled in quadruple precision with the `gfortran` compiler.
The associated type in the header file to call the routines from C is `__float128` and is equivalent to `Float128` (implemented in `Quadmath.jl`).

It will be a nice feature to add an option to generate Julia wrappers for this precision. We just need to add `using Quadmath` at the top of the generated file if Julia wrappers can be generated.

I have a package where I need to write these additional wrappers by hand:
- https://github.com/JuliaSmoothOptimizers/CUTEst.jl/tree/main/gen
- https://github.com/JuliaSmoothOptimizers/CUTEst.jl/blob/main/src/libcutest.jl

**Comment by Gnimuc** [comment]
have you tried `@add_def __float128` before creating the context?

**Reply by Gnimuc** [reply]
then you can define `const __float128 = Float128` in your code before loading the generated code.

**Reply by amontoison** [reply]
I tried `@add_def __float128` before creating the context, and as before, I don't have any wrapper with `__float128` that is generated.
I have the following example of what I would like to generate from the header:
```h
typedef __float128  quadreal;

void cutest_ufn_q_(integer *status, const integer *n, const quadreal *x, quadreal *f);
```
```julia
function cutest_ufn_q_(status, n, x, f)
  @ccall libcutest.cutest_ufn_q_(status::Ptr{Cint}, n::Ptr{Cint}, x::Ptr{__float128}, f::Ptr{__float128})::Cvoid
end
```

**Reply by Gnimuc** [reply]
looks like we need to uncomment this line

https://github.com/JuliaInterop/Clang.jl/blob/9c77fd6b4982c8b34c21a64748aea2664409206e/src/generator/jltypes.jl#L145

**Reply by amontoison** [reply]
The Julia type for `__float128` is `Cfloat128`:
https://github.com/JuliaMath/Quadmath.jl/blob/master/src/Quadmath.jl#L82

`Float128` is working with Fortran routines because we pass everything by reference but if we pass by value, `Cfloat128` is needed.

The macro `@ccall` is also modified if a `Cfloat128` is returned:
https://github.com/JuliaMath/Quadmath.jl/blob/master/src/Quadmath.jl#L68

So it should be safe to handle all headers with `__float128` in `Clang.jl` if the routines don't return a `__float128`.

**Reply by Gnimuc** [reply]
is there a general agreement of defining `Cfloat128` as `NTuple{2,VecElement{Float64}}`, but not a primitive type or an alternative reprensentation`NTuple{2,Float64}`, etc?

if not, package authors should overload `translate(jlty:: JuliaCfloat128, options=Dict()) = :TheirCfloat128Symbol` to control what should be generated.

**Reply by amontoison** [reply]
@ViralBShah Should we add a definition of `Float128` / `Cfloat128` like `Float16` in the Julia language directly so that we have a general agreement on how we should represent this type in Julia?

**Reply by ViralBShah** [reply]
It may be worth bringing up on discourse. For example with BFloat16s.jl, the preference has been to keep it outside to allow for flexibility.
